feat: Add support for Debian versioning scheme

milux opened the pull request:
- add support for Debian versioning scheme
- fix/enhance dynamic range functionality

milux committed:
feat: Add support for Debian versioning scheme, enhance dynamic range functionality

milux committed:
feat: Add support for Debian versioning scheme, enhance dynamic range functionality

tschmidtb51 commented on src/constraint.rs:
```diff
@@ -93,7 +93,13 @@ impl<V: VersionType> VersionConstraint<V> {
+        // i.e., implement a from_native() function for each scheme
+        let (comparator, version) = if constraint_str.starts_with("<<") {
+            (Comparator::LessThan, &constraint_str[2..])
+        } else if constraint_str.starts_with(">>") {
```
Same here with `>>`.

tschmidtb51 commented on src/lib.rs:
```diff
@@ -291,6 +291,34 @@ mod tests {
 
+    #[test]
+    fn test_dynamic_parse_deb() {
+        let range: DynamicVersionRange = "vers:deb/<<1.0".parse().unwrap();
```
This is IMHO not a valid vers as it uses `<<`.

tschmidtb51 commented on src/lib.rs:
```diff
@@ -291,6 +291,34 @@ mod tests {
+        assert!(contains(&range1, "1.0~beta".to_string()).unwrap());
+
+        // 1:1.0 > 2.0 because epoch 1 > 0
+        let range2: DynamicVersionRange = "vers:deb/>>2.0".parse().unwrap();
```
And here.

milux committed:
Fix formatting and clean up

milux committed:
Fix formatting and clean up

milux committed:
feat: Add native constraint parsing via `NativeConstraintConverter`, extend support for Debian and SemVer schemes

milux committed:
refactor: Rename `NativeConstraintConverter` to `NativeVersionConverter`, update related methods and references

milux committed:
refactor: Extract and reorganize test modules for better clarity and maintainability

milux committed:
refactor: Fix formatting for CI pipeline

milux committed:
Minor formatting fix

oxisto commented on src/constraint.rs:
```diff
@@ -7,17 +7,92 @@
+    /// The default implementation calls [`Self::from_native`] and wraps
+    /// the result with [`Self::SCHEME_NAME`]. Schemes whose native syntax
+    /// requires special handling can override this directly.
+    fn from_native_string(raw: &str) -> Result<VersVersionRange<Self>, VersError> {
```
When do we need to call `normalize_and_validate` and when not? Seems the `FromStr` trait of `VersVersonRange` does.

peinjoh replied:
Same question, I think normalize_and_validate should run irrespective of if the schema was supplied directly or autodetected. Currenlty, it will only run as part of parsing that includes `vers:<scheme>`, and not the newly added parse_native API.

milux replied:
I think there was a fundamental decision after that code was initially created. If I recall it correctly, we do not want to normalize anymore **at all**, but instead only validate for correct ordering and raise an error in case a vers string is wrongly ordered. For the time being, I'm going to fix it as you suggested, though...

peinjoh replied:
Then we should document that decision (and follow through with it). Do you have some link for this? Otherwise, this fixes my comment of the checks being asymetrical, so im resolving this for now. We should consider adding an issue to implement the decision mentioned above.

milux replied:
See #41

oxisto commented on src/range/dynamic.rs:
````diff
@@ -27,18 +40,85 @@ use std::str::FromStr;
+    /// ```
+    pub fn parse_native(scheme: &str, raw: &str) -> Result<Self, VersError> {
+        let inner = match scheme {
+            "semver" | "npm" => DynamicVersionRangeInner::SemVer(SemVer::from_native_string(raw)?),
````
I think the "npm" label gets lost here, dont we need to supply that to the version range inner?

peinjoh replied:
It does, the semver / npm schemes all come out with semver, as this is the `SCHEME_NAME` provided for both. This breaks roundtrip parsing.

milux replied:
Right, impls have to carry that along. Fix on the way.

peinjoh replied:
Looking good!

oxisto commented on src/schemes/deb.rs:
```diff
@@ -0,0 +1,290 @@
+    /// as native comparison operators. This method converts them to the
+    /// standard `>` and `<` operators and delegates to the vers parser.
+    fn from_native_constraint(raw: &str) -> Result<VersionConstraint<Self>, VersError> {
+        let converted = raw.replace(">>", ">").replace("<<", "<");
```
what about `=`, isn't that Debian only?

peinjoh replied:
I believe this touches https://github.com/csaf-rs/vers-rs/issues/15 and https://github.com/oasis-tcs/csaf/issues/1445 (Issue is on vls, but there might be a side effect due to the vls <> vers similarity).

As long as #15 is not fixed, I believe this code will handle debian version constraints like `=1.0.0` wrong.

peinjoh replied:
I believe this also allows too much leeway in the debian specific input. Consider the following, invalid debian versions string: `>>=1.0.0` `<<=1.0.0`. These would get turned into `>=1.0` / `<=1.0` respectivly, turning invalid debian versions into valid vers versions. Also `!=`, `<` and `>` are not valid debian version comparators, and would similarly be turned into valid vers version constraints.

milux replied:
Already had the subtile feeling that just doing text replacements might not be sufficient here. Thanks for the counter-example pointing that mistake out. Patch incoming...

peinjoh replied:
Looking good, I imagine we'll have to do this "kind" of constraint parsing for every new versioning scheme now 👍

oxisto commented on src/range/dynamic.rs:
```diff
@@ -27,18 +40,85 @@ use std::str::FromStr;
 #[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
-pub enum DynamicVersionRange {
-    /// SemVer-based range (for "semver" and "npm" schemes)
-    SemVer(GenericVersionRange<SemVer>),
```
I think the README still has GenericVersionRange

oxisto commented on src/schemes/deb.rs:
```diff
@@ -0,0 +1,290 @@
+/// - Upstream version vs. Debian revision separated by last '-'
+/// - Tilde '~' sorts before the end and before any other character
+/// - Sequences of digits are compared numerically; non-digits lexicographically
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
```
I think you need to drop PartialEq and Eq from the derive and implement this yourself, otherwise there might be inconsistencies petween partial ordering and equality because of the Debian revision "fix" in cmp (to set it "0" if its empty)


```rust
impl PartialEq for DebVersion {
      fn eq(&self, other: &Self) -> bool {
          self.cmp(other) == Ordering::Equal
      }
  }
  ```

peinjoh replied:
Looking good, IMO, @oxisto please reopen if you see that this still needs more work done.

milux committed:
feat: Improve Debian versioning support and dynamic range handling

- Add strict parsing and validation for Debian native constraints.
- Enhance `parse_native` to maintain correct versioning scheme labels.
- Implement normalization and validation for parsed ranges.
- Update `from_native_string` to support scheme-specific handling.
- Extend tests to cover new scenarios and edge cases.
